### PR TITLE
Support import of raw cca tokens with C_ObjectCreate

### PR DESCRIPTION
**Pull request by hfreude**
3 commits:
- TESTCASES: new testcase to test CCA opaque key token export and import
- CCA: support import of raw cca tokens with C_ObjectCreate via CKA_IBM_OPAQUE
- CCA: store only public token in IBM_OPAQUE attribute
for more details pleas see header of each commit

### CODE REVIEW

**Comment by hfreude** [comment]
PR is updated, now without AES cipher key support and without requirement for CKA_VALUE for des and aes keys.

File: usr/lib/common/new_host.c
Comment (ifranzki) [inline usr/lib/common/new_host.c:3641]: Please check if this has to be corrected in the other copies of new_host.c as well (ep11 token and icsf token).

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:245]: In ep11_specific.c there is a macro defined to dump memory:
`#define TRACE_DEBUG_DUMP(_buf, _buflen) hexdump(_buf, _buflen)`
Function hexdump() is alos local there. 
Maybe we can add a function into trace.c and a macro like TRACE_DEBUG_DUMP to use it, so that it can be used everywhere?

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:271]: This should probably go into attributes.c or p11util.c so that it is available for others, too. 
In p11utilc, there are already functions p11_get_ckr() and p11_get_ckm() to get the textual representation of return codes and mechanisms. maybe it would make sense to add a p11_get_cka() as well?

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:364]: move to template.c, or p11util.c ?

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:399]: I usually like struct to overlay over the buffer better than hard coded offsets, but since the CCA token uses the offset-way already, I am fine with this, too.

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:419]: 2 levels of switch()-es would be easier to read maybe:
```
switch (t[0]) {
case 0x01:
    switch (t[4]) {
    case xxx:
    }
}
```

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:5060]: Actually, you could check here already if an CKA_IBM_OPAQUE is there, and if so, then analyze it and find out about the key type and class from the token. This would lift the requirement for the user to specify CLA_KEY_TYPE and CKA_CLASS when importing an opaque key. The key token does in fact contain all this information already, so no need for the user to specify it.
The only question would be if an EP11 key token also allows to get the same kind of information from it. So once we do the key blob import for EP11, it then should behave the same as for CCA.

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:5255]: If you add this here (which makes sense, now that we have this functions), why not also adding it to the other keygen functions as well?
Any maybe adjust the trace message, here is is not an 'imported' key, but a 'generated' one.

File: testcases/misc_tests/cca_export_import_test.c
Comment (ifranzki) [inline testcases/misc_tests/cca_export_import_test.c:1]: I guess we should have the standard file header here with copyright infos, license, etc.

File: usr/lib/common/new_host.c
Comment (hfreude) [reply usr/lib/common/new_host.c:3641]: needs also to get fixed in the other new_host.c files ...
I'll do that - maybe split these fixes into a new commit.

File: usr/lib/cca_stdll/cca_specific.c
Comment (hfreude) [reply usr/lib/cca_stdll/cca_specific.c:245]: yes, makes sense - will do this

File: usr/lib/cca_stdll/cca_specific.c
Comment (hfreude) [reply usr/lib/cca_stdll/cca_specific.c:399]: I have the structs - but as most of the cca specific code deals with some kind of offsets and would need a total rework to adapt to structs, I've chosen to use these simple offset values.

File: usr/lib/cca_stdll/cca_specific.c
Comment (hfreude) [reply usr/lib/cca_stdll/cca_specific.c:5060]: I'll leave this to Reinhard to decide. It's doable but then a user does not need and provide any key information but only the CCA secure key token. So no way to check if this was the key type the user intended to import.

File: testcases/misc_tests/cca_export_import_test.c
Comment (hfreude) [reply testcases/misc_tests/cca_export_import_test.c:1]: Ok, will do.

File: usr/lib/cca_stdll/cca_specific.c
Comment (hfreude) [reply usr/lib/cca_stdll/cca_specific.c:364]: done, will be part of the next PR update

File: usr/lib/cca_stdll/cca_specific.c
Comment (hfreude) [reply usr/lib/cca_stdll/cca_specific.c:419]: well, no
the idea of the whole function is to group it based on keytypes. Mixing of ifs and switches and more levels may give the function a 1us performance boost but the human readability is much more worse. At least this is my understanding.

**Comment by hfreude** [comment]
Ok, hopefully this was the last update of the PR:
- added some more attributes for the dump_attr function.
- added CKA_SENSITIVE=true for the import of sym and priv keys (request from Reinhard)

File: usr/lib/cca_stdll/cca_specific.c
Comment (ifranzki) [inline usr/lib/cca_stdll/cca_specific.c:403]: We could move this function to common code (e.g. utility.c, were build_attribute() is already) so that others could also utilize it, since many other places also have the sequence of build_attribute() followed by template_update_attribute(), but that's not something that should hold off this PR.

File: testcases/misc_tests/cca_export_import_test.c
Comment (ifranzki) [inline testcases/misc_tests/cca_export_import_test.c:27]: I don't find this define anywhere else except in this testcase, especially not in cca_specific.c. Is this a left over, or did you leave it there intentionally?

**Review by ifranzki — APPROVED** [review]
lgtm

File: testcases/misc_tests/cca_export_import_test.c
Comment (hfreude) [reply testcases/misc_tests/cca_export_import_test.c:27]: You are right. This is local to the file.
The code to support import of CCA AES cipher blobs is easy - just add it to the switch within the symetric key import code function. However, I did not want to discard the test code as I spend some time to develop it and so I commented it out with this #ifdef CCA_AES_CIPHER_KEY_SUPPORTED

File: usr/lib/cca_stdll/cca_specific.c
Comment (hfreude) [reply usr/lib/cca_stdll/cca_specific.c:403]: yes, this could and should be done. And - as this function avoids the free() if update_attribute() fails, all the places where build_attribute() + update_attribute() is used would benefit from this.

**Review by p-steuer — APPROVED** [review]
lgtm

**Pull request merged**
